modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Environment.scala

```diff
@@ -204,7 +204,7 @@ object Environment {
       getRegionFromConfig(file).orElse(getRegion),
       cloud,
       acceptInvalid,
-      Some(() => metadata.submit.compile.drain)
+      Some(() => metadata.submit.interruptAfter(1.minute).compile.drain)
```

Comment:
1. Why is `preShutdown` an `Option` ? Is it not always `Some` ?
2. In the integration test, I doubt that the issue was that `preShutdown` was taking longer than 1 minute, it would rather seem that it would get stuck forever
3. If I understand correctly, the reason why we need this `preShutdown` is that unlike all the other streams [here](https://github.com/snowplow/enrich/blob/master/modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Run.scala#L151-L153), we want to flush the buffer of metadata before exiting. Ideally, we would want that to be automatically triggered when [enrich](https://github.com/snowplow/enrich/blob/master/modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Run.scala#L150) stream terminates. What about using `enrich.onComplete()` to flush the metadata ? This way you wouldn't need to have this `preShutdown` in the `Environment` and to worry about it. The flushing needs to be a function and not a stream.

Reply:
That's a really good architectural point!